File: ext/RmlUi.Native/src/SystemInterface.cpp
#include "RmlNative.h"
#include "RmlUi/Core/SystemInterface.h"
#include <iostream>
#include <string>

typedef double(*onGetElapsedTime)();
typedef const char *(*onTranslateString)(int *changeCount, const char *input);
typedef bool(*onLogMessage)(Rml::Log::Type type, const char *message);
typedef char*(*onJoinPath)(const char *document_path, const char *path);
typedef bool(*onSetMouseCursor)(const char *message);
typedef void(*onSetClipboardText)(const char *message);
typedef const char* (*onGetClipboardText)();
typedef void(*onActivateKeyboard)(float caret_x, float caret_y, float line_height);
typedef void(*onDeactivateKeyboard)();

class SystemInterface : Rml::SystemInterface {
private:
    ::onGetElapsedTime m_onGetElapsedTime;
    ::onTranslateString m_onTranslateString;
    ::onLogMessage m_onLogMessage;
    ::onJoinPath m_onJoinPath;
    ::onSetMouseCursor m_onSetMouseCursor;
    ::onSetClipboardText m_onSetClipboardText;
    ::onGetClipboardText m_onGetClipboardText;
    ::onActivateKeyboard m_onActivateKeyboard;
    ::onDeactivateKeyboard m_onDeactivateKeyboard;

public:
    explicit SystemInterface(::onGetElapsedTime onGetElapsedTime, ::onTranslateString onTranslateString,
                             ::onLogMessage onLogMessage, ::onJoinPath onJoinPath, ::onSetMouseCursor onSetMouseCursor,
                             ::onSetClipboardText onSetClipboardText, ::onGetClipboardText onGetClipboardText,
                             ::onActivateKeyboard onActivateKeyboard, ::onDeactivateKeyboard onDeactivateKeyboard) {
        m_onGetElapsedTime = onGetElapsedTime;
        m_onTranslateString = onTranslateString;
        m_onLogMessage = onLogMessage;
        m_onJoinPath = onJoinPath;
        m_onSetMouseCursor = onSetMouseCursor;
        m_onSetClipboardText = onSetClipboardText;
        m_onGetClipboardText = onGetClipboardText;
        m_onActivateKeyboard = onActivateKeyboard;
        m_onDeactivateKeyboard = onDeactivateKeyboard;
    }

    double GetElapsedTime() override {
        return (*m_onGetElapsedTime)();
    }

    int TranslateString(Rml::String &translated, const Rml::String &input) override {
        int changeCount = 0;
        //translated = (*m_onTranslateString)(&changeCount, input.c_str());
        translated = input;

        return changeCount;
    }

    bool LogMessage(Rml::Log::Type type, const Rml::String &message) override {
        return (*m_onLogMessage)(type, message.c_str());
    }

    void JoinPath(Rml::String& translated_path, const Rml::String& document_path, const Rml::String& path) {
        translated_path = (*m_onJoinPath)(document_path.c_str(), path.c_str());
    }

    void SetMouseCursor(const Rml::String &cursor_name) override {
        (*m_onSetMouseCursor)(cursor_name.c_str());
    }

    void SetClipboardText(const Rml::String &text) override {
        (*m_onSetClipboardText)(text.c_str());
    }

    void GetClipboardText(Rml::String &text) override {
        text = (*m_onGetClipboardText)();
    }

    void ActivateKeyboard(Rml::Vector2f caret_position, float line_height) override {
        (*m_onActivateKeyboard)(caret_position.x, caret_position.y, line_height);
    }

    void DeactivateKeyboard() override {
        (*m_onDeactivateKeyboard)();
    }
};

RMLUI_CAPI void *rml_SystemInterface_New(::onGetElapsedTime onGetElapsedTime, ::onTranslateString onTranslateString,
                                         ::onLogMessage onLogMessage, ::onJoinPath onJoinPath, ::onSetMouseCursor onSetMouseCursor,
                                         ::onSetClipboardText onSetClipboardText, ::onGetClipboardText onGetClipboardText,
                                         ::onActivateKeyboard onActivateKeyboard, ::onDeactivateKeyboard onDeactivateKeyboard) {
    return new SystemInterface(onGetElapsedTime, onTranslateString, onLogMessage, onJoinPath, onSetMouseCursor,
            onSetClipboardText, onGetClipboardText, onActivateKeyboard, onDeactivateKeyboard);
}

File: ext/RmlUI.Net/SystemInterface.cs
using System;
using System.Text;
using static RmlUiNet.Native.SystemInterface;

namespace RmlUiNet
{
    public abstract class SystemInterface : RmlBase<SystemInterface>
    {
        private Native.SystemInterface.OnGetElapsedTime _onGetElapsedTime;
        private Native.SystemInterface.OnTranslateString _onTranslateString;
        private Native.SystemInterface.OnLogMessage _onLogMessage;
        private Native.SystemInterface.OnJoinPath _onJoinPath;
        private Native.SystemInterface.OnSetMouseCursor _onSetMouseCursor;
        private Native.SystemInterface.OnSetClipboardText _onSetClipboardText;
        private Native.SystemInterface.OnGetClipboardText _onGetClipboardText;
        private Native.SystemInterface.OnActivateKeyboard _onActivateKeyboard;
        private Native.SystemInterface.OnDeactivateKeyboard _onDeactivateKeyboard;

        public SystemInterface() : base(IntPtr.Zero)
        {
            _onGetElapsedTime = OnGetElapsedTimeInternal;
            _onTranslateString = OnTranslateStringInternal;
            _onLogMessage = LogMessage;
            _onJoinPath = JoinPath;
            _onSetMouseCursor = OnSetMouseCursor;
            _onSetClipboardText = OnSetClipboardText;
            _onGetClipboardText = OnGetClipboardText;
            _onActivateKeyboard = OnActivateKeyboard;
            _onDeactivateKeyboard = OnDeactivateKeyboard;

            NativePtr = Native.SystemInterface.Create(
                _onGetElapsedTime,
                _onTranslateString,
                _onLogMessage,
                _onJoinPath,
                _onSetMouseCursor,
                _onSetClipboardText,
                _onGetClipboardText,
                _onActivateKeyboard,
                _onDeactivateKeyboard
            );

            ManuallyRegisterCache(NativePtr, this);
        }

        public virtual string JoinPath(string path, string file) {
            return System.IO.Path.Combine(path, file);
        }

        private double OnGetElapsedTimeInternal()
        {
            return ElapsedTime;
        }

        private string OnTranslateStringInternal(ref int changeCount, string input)
        {
            return TranslateString(out changeCount, input);
        }

        private void OnSetMouseCursor(string mouseCursor)
        {
            SetMouseCursor(mouseCursor);
        }

        /// <summary>
        /// Get the number of seconds elapsed since the start of the application.
        /// </summary>
        public abstract double ElapsedTime { get; }

        /// <summary>
        /// Translate the input string into the translated string.
        /// </summary>
        /// <param name="changeCount">Number of translations that occurred.</param>
        /// <param name="input">String as received from XML.</param>
        /// <returns>Translated string ready for display.</returns>
        public virtual string TranslateString(out int changeCount, string input)
        {
            changeCount = 0;
            return input;
        }

        public virtual bool LogMessage(LogType type, string message)
        {
            Console.WriteLine("[{0}] {1}", type, message);

            return true;
        }

        public virtual void SetMouseCursor(string cursor) {

        }

        private void OnSetClipboardText(string text)
        {
            SetClipboardText(text);
        }

        public abstract void SetClipboardText(string text);

        private string OnGetClipboardText()
        {
            return GetClipboardText();
        }

        public abstract string GetClipboardText();

        private void OnActivateKeyboard(float caretX, float caretY, float lineHeight)
        {
            ActivateKeyboard(caretX, caretY, lineHeight);
        }

        public abstract void ActivateKeyboard(float caretX, float caretY, float lineHeight);

        private void OnDeactivateKeyboard()
        {
            DeactivateKeyboard();
        }

        public abstract void DeactivateKeyboard();
    }

    public enum LogType
    {
        Always = 0,
        Error,
        Assert,
        Warning,
        Info,
        Debug,
    }
}
